src/unity/toolkits/tcmps/mps_graph_layers.mm

```diff
@@ -98,22 +98,22 @@
   [weight setLearningRate:lr];
 }
 
-void ConvGraphLayer::Export(
-    std::unordered_map<std::string,
-                       std::tuple<std::string, float *, int, std::vector<int>>>
-        &table) {
-  int k_h = iparams[0];
-  int k_w = iparams[1];
-  int c_in = iparams[2];
-  int c_out = iparams[3];
+float_array_map ConvGraphLayer::Export() const {
+  float_array_map table;
+  size_t k_h = iparams[0];
+  size_t k_w = iparams[1];
+  size_t c_in = iparams[2];
+  size_t c_out = iparams[3];
   [weight load];
   std::string weight_key = name + "_weight";
-  table[weight_key] = {
-      weight_key, (float *)[weight weights], 4, {c_out, k_h, k_w, c_in}};
+  table[weight_key] = shared_float_array::copy(
+      reinterpret_cast<float*>([weight weights]), {c_out, k_h, k_w, c_in});
```

Comment:
Good question. This will actually create a new `shared_ptr` to a `float_array_buffer`, and _copies_ all the data from `weight` into that newly allocated buffer. So yes, this code is actually less efficient than the old code, insofar as it incurs another copy. But I really wanted to stop exposing the client code (in this case, in Python!) to memory addresses inside of these Objective C / MPS objects. Especially since the data in these internal objects is mutable, and will change after the next training iteration. So now the export methods actually create a standalone, immutable snapshot of the weights, instead of a map from parameter names to mutable weights owned by TCMS.

Reply:
Got it! That makes sense. Thanks for the clarification!